runtime/ftplugin/help.vim

```diff
@@ -21,77 +21,5 @@ endif
 " Prefer Vim help instead of manpages.
 setlocal keywordprg=:help
 
-if !exists('g:no_plugin_maps')
-  function! s:show_toc() abort
-    let bufname = bufname('%')
-    let info = getloclist(0, {'winid': 1})
-    if !empty(info) && getwinvar(info.winid, 'qf_toc') ==# bufname
-      lopen
-      return
-    endif
-
-    let toc = []
-    let lnum = 2
-    let last_line = line('$') - 1
-    let last_added = 0
-    let has_section = 0
-    let has_sub_section = 0
-
-    while lnum && lnum <= last_line
-      let level = 0
-      let add_text = ''
-      let text = getline(lnum)
-
-      if text =~# '^=\+$' && lnum + 1 < last_line
-        " A de-facto section heading.  Other headings are inferred.
-        let has_section = 1
-        let has_sub_section = 0
-        let lnum = nextnonblank(lnum + 1)
-        let text = getline(lnum)
-        let add_text = text
-        while add_text =~# '\*[^*]\+\*\s*$'
-          let add_text = matchstr(add_text, '.*\ze\*[^*]\+\*\s*$')
-        endwhile
-      elseif text =~# '^[A-Z0-9][-A-ZA-Z0-9 .][-A-Z0-9 .():]*\%([ \t]\+\*.\+\*\)\?$'
-        " Any line that's yelling is important.
-        let has_sub_section = 1
-        let level = has_section
-        let add_text = matchstr(text, '.\{-}\ze\s*\%([ \t]\+\*.\+\*\)\?$')
-      elseif text =~# '\~$'
-            \ && matchstr(text, '^\s*\zs.\{-}\ze\s*\~$') !~# '\t\|\s\{2,}'
-            \ && getline(lnum - 1) =~# '^\s*<\?$\|^\s*\*.*\*$'
-            \ && getline(lnum + 1) =~# '^\s*>\?$\|^\s*\*.*\*$'
-        " These lines could be headers or code examples.  We only want the
-        " ones that have subsequent lines at the same indent or more.
```

Comment:
And this is why we can't have nice things... 😞 

It's impossible to replace this sort of hackery with syntax tree parsing unless we take structured documentation a lot more seriously.

Reply:
What part of `text =~# '\~$ && matchstr(text, '^\s*\zs.\{-}\ze\s*\~$') !~# '\t\|\s\{2,} && getline(lnum - 1) =~# '^\s*<\?$\|^\s*\*.*\*$' && getline(lnum + 1) =~# '^\s*>\?$\|^\s*\*.*\*$'` do you find hacky?

Reply:
The fact that the function is building its own heading toc hierarchy based on heuristics instead of relying on syntax. (Understandably, given the way docs are written.)

Reply:
Why is it "impossible"? `gen_help_html.lua` has workarounds by inspecting the text content.

Reply:
Yes, and this sort of "check the contents to see if it's _really_ a heading" is exactly what I want to replace with just being able to rely on a proper syntax tree. (I wrote "impossible to replace", not "impossible to do".)